mesonbuild/compilers/cpp.py

```diff
@@ -0,0 +1,199 @@
+            extra_args = []
+        fargs = {'prefix': prefix, 'header': hname, 'symbol': symbol}
+        t = '''{prefix}
+        #include <{header}>
+        using {symbol};
+        int main () {{ return 0; }}'''
+        return self.compiles(t.format(**fargs), env, extra_args, dependencies)
+
+
+class VisualStudioCPPCompiler(VisualStudioCCompiler, CPPCompiler):
+    def __init__(self, exelist, version, is_cross, exe_wrap):
+        self.language = 'cpp'
+        CPPCompiler.__init__(self, exelist, version, is_cross, exe_wrap)
+        VisualStudioCCompiler.__init__(self, exelist, version, is_cross, exe_wrap)
+        self.base_options = ['b_pch'] # FIXME add lto, pgo and the like
+
+    def get_options(self):
+        return {'cpp_eh': coredata.UserComboOption('cpp_eh',
+                                                   'C++ exception handling type.',
+                                                   ['none', 'a', 's', 'sc'],
+                                                   'sc'),
+                'cpp_winlibs': coredata.UserStringArrayOption('cpp_winlibs',
+                                                              'Windows libs to link against.',
+                                                              msvc_winlibs)
+                }
+
+    def get_option_compile_args(self, options):
+        args = []
+        std = options['cpp_eh']
+        if std.value != 'none':
+            args.append('/EH' + std.value)
+        return args
+
+    def get_option_link_args(self, options):
+        return options['cpp_winlibs'].value[:]
+
+    def get_compiler_check_args(self):
+        # Visual Studio C++ compiler doesn't support -fpermissive,
+        # so just use the plain C args.
+        return super(VisualStudioCCompiler, self).get_compiler_check_args()
+
+
+class GnuCPPCompiler(GnuCompiler, CPPCompiler):
+    def __init__(self, exelist, version, gcc_type, is_cross, exe_wrap, defines):
+        CPPCompiler.__init__(self, exelist, version, is_cross, exe_wrap)
+        GnuCompiler.__init__(self, gcc_type, defines)
+        default_warn_args = ['-Wall', '-Winvalid-pch', '-Wnon-virtual-dtor']
+        self.warn_args = {'1': default_warn_args,
+                          '2': default_warn_args + ['-Wextra'],
+                          '3': default_warn_args + ['-Wextra', '-Wpedantic']}
+
+    def get_options(self):
+        opts = {'cpp_std': coredata.UserComboOption('cpp_std', 'C++ language standard to use',
+                                                    ['none', 'c++03', 'c++11', 'c++14', 'c++1z',
+                                                     'gnu++03', 'gnu++11', 'gnu++14', 'gnu++1z'],
+                                                    'none'),
+                'cpp_debugstl': coredata.UserBooleanOption('cpp_debugstl',
+                                                           'STL debug mode',
+                                                           False)}
+        if self.gcc_type == GCC_MINGW:
```

Comment:
__[flake8]__
* [F405] 'GCC_MINGW' may be undefined, or defined from star imports: .compilers

<!-- Comment by [SideCI](https://sideci.com) -->